src/ir/ExpressionAnalyzer.cpp

```diff
@@ -459,37 +460,41 @@ HashType ExpressionAnalyzer::hash(Expression* curr) {
         }
         // Sometimes children are optional, e.g. return, so we must hash
         // their number as well.
-        hash(counter);
+        rehash32(digest, counter);
       }
     }
 
-    void hash(HashType hash) { digest = rehash(digest, hash); }
-    void hash64(uint64_t hash) {
-      digest = rehash(rehash(digest, HashType(hash >> 32)), HashType(hash));
-    }
-
     void visitScopeName(Name curr) {
       // Names are relative, we give the same hash for
       // (block $x (br $x))
       // (block $y (br $y))
       static_assert(sizeof(Index) == sizeof(int32_t),
                     "wasm64 will need changes here");
       assert(internalNames.find(curr) != internalNames.end());
-      return hash(internalNames[curr]);
+      rehash32(digest, internalNames[curr]);
+    }
+    void visitNonScopeName(Name curr) { rehash32(digest, uint64_t(curr.str)); }
+    void visitInt(int32_t curr) { rehash32(digest, uint32_t(curr)); }
+    void visitLiteral(Literal curr) {
+      // can't use std::hash overload here
+      uint8_t bytes[16];
+      curr.getBits(bytes);
+      uint64_t chunks[2];
+      memcpy(chunks, bytes, sizeof(chunks));
+      rehash32(digest, curr.type.getID());
+      rehash32(digest, chunks[0]);
+      rehash32(digest, chunks[1]);
     }
-    void visitNonScopeName(Name curr) { return hash64(uint64_t(curr.str)); }
-    void visitInt(int32_t curr) { hash(curr); }
-    void visitLiteral(Literal curr) { hash(std::hash<Literal>()(curr)); }
-    void visitType(Type curr) { hash(int32_t(curr.getSingle())); }
+    void visitType(Type curr) { rehash32(digest, curr.getID()); }
     void visitIndex(Index curr) {
       static_assert(sizeof(Index) == sizeof(int32_t),
                     "wasm64 will need changes here");
-      hash(int32_t(curr));
+      rehash32(digest, curr);
     }
     void visitAddress(Address curr) {
       static_assert(sizeof(Address) == sizeof(int32_t),
                     "wasm64 will need changes here");
-      hash(int32_t(curr));
+      rehash32(digest, curr.addr); // careful when changing, determinism?
```

Comment:
It would be good to make this comment more specific. Be careful when changing what? Why?

Reply:
Hmm, when thinking about it, this comment it probably unnecessary anyway. Was considering the implications of having addresses of different sizes when adapting to Wasm64 as indicated by the assert right above, yet the decision which one to pick would still be deterministic based on the target. Going to remove the comment.